### DISCUSSION TITLE
Run a C++ model from python?

### DISCUSSION THREAD

**Discussion by jacksonsdean** [FLAMEGPU/FLAMEGPU2#927] [Q&A] [answered]
Is there a way to define a model in C++ and then run that model in a Python CUDA ensemble?

We have a model already created in C++, but we want to wrap an evolutionary algorithm around it in Python.

**Comment by Robadob** [comment, marked as answer]
Officially via FLAMEGPU, no. All our Python bindings are generated via [SWIG](https://www.swig.org/). We tell it what we want wrapped in [`swig/python/flamegpu.i`](https://github.com/FLAMEGPU/FLAMEGPU2/blob/master/swig/python/flamegpu.i).

If you wanted to achieve this, you'd probably want to create a C++ method that returns (or fills) a `ModelDescription` with your model (I expect this would need to be part of the main `flamegpu` project. Then extend `swig/python/flamegpu.i`, so that SWIG additionally wraps that method to add it to the Python interface. If all is done correctly, when you rebuild `pyflamegpu` your method should now be available in the `pyflamegpu` module.

It's not something we've tried or required, and SWIG can be a bit unpredictable at times, so I'm not sure if any unexpected issues would occur whilst trying to get SWIG to play nicely.

**Reply by Robadob** [reply]
Theoretically, it should be possible to have SWIG wrap a C++ model in a separate (dependent) Python module. This would be a much cleaner solution that we could officially provide (via a template repository), however also more complicated to setup in the first place.

I've created an issue (https://github.com/FLAMEGPU/FLAMEGPU2/issues/928) so that it's discussed by the team, but I don't expect it will be a priority.

**Comment by mondus** [comment]
Alternatively, translating the model to Python (using C++ style agent functions) should be fairly painless as there is a one to one mapping of all API calls. This obviously depends on if you have done anything with any other C++ libraries. In which case Robs answer is sensible. It would also be possible to have Python drive a C++ ensemble indirectly through file IO.
